Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+  else
+    return r;
+}
+
+/// Returns the green color channel for detector colors
+/// @return green color channel
+std::vector<int> const &StructuredDetector::getG() const {
+  if (m_map)
+    return m_base->getG();
+  else
+    return g;
+}
+
+/// Returns the blue color channel for detector colors
+/// @return blue color channel
+std::vector<int> const &StructuredDetector::getB() const {
+  if (m_map)
+    return m_base->getB();
+  else
+    return b;
+}
+
+/// Returns the number of pixels in the Y direction.
+/// @return number of Y pixels
+int StructuredDetector::ypixels() const {
+  if (m_map)
+    return m_base->m_ypixels;
+  else
+    return this->m_ypixels;
+}
+
+/// Returns the idstart
+int StructuredDetector::idstart() const {
+  if (m_map)
+    return m_base->m_idstart;
+  else
+    return this->m_idstart;
+}
+
+/// Returns the idfillbyfirst_y
+bool StructuredDetector::idfillbyfirst_y() const {
+  if (m_map)
+    return m_base->m_idfillbyfirst_y;
+  else
+    return this->m_idfillbyfirst_y;
+}
+
+/// Returns the idstepbyrow
+int StructuredDetector::idstepbyrow() const {
+  if (m_map)
+    return m_base->m_idstepbyrow;
+  else
+    return this->m_idstepbyrow;
+}
+
+/// Returns the idstep
+int StructuredDetector::idstep() const {
+  if (m_map)
+    return m_base->m_idstep;
+  else
+    return this->m_idstep;
+}
+
+std::vector<double> const &StructuredDetector::getXValues() const {
+  if (m_map)
+    return m_base->getXValues();
+  else
+    return this->m_xvalues;
+}
+
+std::vector<double> const &StructuredDetector::getYValues() const {
+  if (m_map)
+    return m_base->getYValues();
+  else
+    return this->m_yvalues;
+}
+
+/** Initialize a StructuredDetector by creating all of the pixels
+* contained within it. You should have set the name, position
+* and rotation and facing of this object BEFORE calling this.
+* NB xpixels and ypixels requires (xpixels+1)*(ypixels+1) vertices
+*
+* @param xpixels :: number of pixels in X
+* @param ypixels :: number of pixels in Y
+* @param x :: X vertices
+* @param y :: Y vertices
+* @param idstart :: detector ID of the first pixel
+* @param idfillbyfirst_y :: set to true if ID numbers increase with Y indices
+*first. That is: (0,0)=0; (0,1)=1, (0,2)=2 and so on.
+* @param idstepbyrow :: amount to increase the ID number on each row. e.g, if
+*you fill by Y first,
+*            and set  idstepbyrow = 100, and have 50 Y pixels, you would get:
+*            (0,0)=0; (0,1)=1; ... (0,49)=49; (1,0)=100; (1,1)=101; etc.
+* @param idstep :: amount to increase each individual ID number with a row.
+*e.g, if you fill by Y first,
+*            and idstep=100 and idstart=1 then (0,0)=1; (0,1)=101; and so on
+*
+*/
+void StructuredDetector::initialize(int xpixels, int ypixels,
+                                    const std::vector<double> &x,
+                                    const std::vector<double> &y, int idstart,
+                                    bool idfillbyfirst_y, int idstepbyrow,
```

Comment:
idstepbyrow is fixed, so I think that answers my previous question about whether you could have x_ids as a function of y. I think that just needs to be made clear in the documentation then that this is a kind of contorted rectangular grid.